Missing Makefile Dependency

## Bug Description
<!--A clear and concise description of the problem (Note: A missing feature is not a bug).-->

We have started recently seeing failures in our build system where header files are not found during compilation. We do not have a clear indicator as to when this started occuring, but we are seeing it somewhat regularly on the build systems over the past several weeks (June 2023). 

## Steps to Reproduce
<!--Steps to reproduce the behavior (input file, or modifications to an existing input file, etc.)-->

Reproducing is a nightmare, but I after putting sleep statements in some of the symlinking logic and printing the actual linking, I noticed the following:

```
permcj@{~/projects/moose/test}[next]$ make -j10                                                                                                                                            (moose)
ln -sf /Users/permcj/projects/moose/test/include/kernels/NanKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/NanKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/ExampleShapeElementKernel2.h /Users/permcj/projects/moose/test/build/header_symlinks/ExampleShapeElementKernel2.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/CoupledEigenKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/CoupledEigenKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/ExceptionKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ExceptionKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DeprecatedParamKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DeprecatedParamKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/PMassEigenKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/PMassEigenKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/NonlocalShapeElementKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/NonlocalShapeElementKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/MaterialEigenKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/MaterialEigenKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/SimpleTestShapeElementKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/SimpleTestShapeElementKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DoNotCopyParametersKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DoNotCopyParametersKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/ExpiredKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ExpiredKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/TestADScalarKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/TestADScalarKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/MatKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/MatKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/PMassKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/PMassKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/ExampleShapeElementKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ExampleShapeElementKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DiffTensorKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DiffTensorKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/MultipleUpdateErrorKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/MultipleUpdateErrorKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/RenameCoupledScalarVarScalarKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/RenameCoupledScalarVarScalarKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/NanAtCountKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/NanAtCountKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/RenamedKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/RenamedKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DeprecatedKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DeprecatedKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/PhiZeroKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/PhiZeroKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DefaultMatPropConsumerKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DefaultMatPropConsumerKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DotCouplingKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DotCouplingKernel.h
ln -sf /Users/permcj/projects/moose/test/include/scalarkernels/PostprocessorSinkScalarKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/PostprocessorSinkScalarKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DiffMKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DiffMKernel.h
ln -sf /Users/permcj/projects/moose/test/include/scalarkernels/VectorPostprocessorScalarKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/VectorPostprocessorScalarKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/DeprecatedCoupledVarKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/DeprecatedCoupledVarKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/ConsoleMessageKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ConsoleMessageKernel.h
ln -sf /Users/permcj/projects/moose/test/include/kernels/UserObjectKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/UserObjectKernel.h
ln -sf /Users/permcj/projects/moose/test/include/fvkernels/FVFunctorElementalKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/FVFunctorElementalKernel.h
ln -sf /Users/permcj/projects/moose/test/include/nodalkernels/ADLowerBoundNodalKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ADLowerBoundNodalKernel.h
ln -sf /Users/permcj/projects/moose/test/include/nodalkernels/CoefTimeDerivativeNodalKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/CoefTimeDerivativeNodalKernel.h
ln -sf /Users/permcj/projects/moose/test/include/nodalkernels/ADUpperBoundNodalKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ADUpperBoundNodalKernel.h
ln -sf /Users/permcj/projects/moose/test/include/actions/AddMatAndKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/AddMatAndKernel.h
ln -sf /Users/permcj/projects/moose/test/include/actions/BadAddKernelAction.h /Users/permcj/projects/moose/test/build/header_symlinks/BadAddKernelAction.h
ln -sf /Users/permcj/projects/moose/test/include/dgkernels/MatDGKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/MatDGKernel.h
ln -sf /Users/permcj/projects/moose/test/include/constraints/GapHeatConductanceAuxKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/GapHeatConductanceAuxKernel.h
ln -sf /Users/permcj/projects/moose/test/include/userobjects/TestADScalarKernelUserObject.h /Users/permcj/projects/moose/test/build/header_symlinks/TestADScalarKernelUserObject.h
ln -sf /Users/permcj/projects/moose/test/include/bcs/CoupledKernelGradBC.h /Users/permcj/projects/moose/test/build/header_symlinks/CoupledKernelGradBC.h
Creating Unity /Users/permcj/projects/moose/test/build/unity_src/kernels_Unity.C
Compiling C++ (in opt mode) /Users/permcj/projects/moose/test/build/unity_src/kernels_Unity.C...
ln -sf /Users/permcj/projects/moose/test/include/nodalkernels/ADCoupledForceNodalKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ADCoupledForceNodalKernel.h
Linking Library /Users/permcj/projects/moose/test/lib/libmoose_test-opt.la...
Linking Executable /Users/permcj/projects/moose/test/moose_test-opt...
```

Specifically, the ordering here:
```
Compiling C++ (in opt mode) /Users/permcj/projects/moose/test/build/unity_src/kernels_Unity.C...
ln -sf /Users/permcj/projects/moose/test/include/nodalkernels/ADCoupledForceNodalKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ADCoupledForceNodalKernel.h
```

While this didn't trigger an error, I was able to remove and touch other files and see a few of these linking statements appearing after compilation steps. However, I was never able to get any of these to appear out of order on the system that they would normally appear in (e.g. Kernels linking after starting compilation of the Kernels unity file), however, after looking at the Makefile, I couldn't rule out that this wasn't an issue. After including one of these file that seemed to be linking "late" on my local system in an unrelated file, I was able to reproduce the "file not found error"

Example:

```
Creating Unity /Users/permcj/projects/moose/test/build/unity_src/kernels_Unity.C
Compiling C++ (in opt mode) /Users/permcj/projects/moose/test/build/unity_src/kernels_Unity.C...
In file included from /Users/permcj/projects/moose/test/build/unity_src/kernels_Unity.C:41:
/Users/permcj/projects/moose/test/src/kernels/ExceptionKernel.C:13:10: fatal error: 'ADCoupledForceNodalKernel.h' file not found
#include "ADCoupledForceNodalKernel.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [/Users/permcj/projects/moose/framework/build.mk:149: /Users/permcj/projects/moose/test/build/unity_src/kernels_Unity.aarch64-apple-darwin20.0.0.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
ln -sf /Users/permcj/projects/moose/test/include/nodalkernels/ADCoupledForceNodalKernel.h /Users/permcj/projects/moose/test/build/header_symlinks/ADCoupledForceNodalKernel.h
```

## Impact
<!--Does this prevent you from getting your work done, or is it more of an annoyance?-->
Build systems with missing dependencies are very frustrating as they cause users and developers alike to search for issues that they likely did not create. Major annoyance

--- a/framework/app.mk
+++ b/framework/app.mk
@@ -144,20 +144,6 @@ test_fobjects:= $(patsubst %.f, %.$(obj-suffix), $(test_fsrcfiles))
 test_f90objects:= $(patsubst %.f90, %.$(obj-suffix), $(test_f90srcfiles))
 app_test_objects := $(test_objects) $(test_cobjects) $(test_fobjects) $(test_f90objects)
 
-ifeq ($(MOOSE_HEADER_SYMLINKS),true)
-
-$(app_objects): $(moose_config_symlink)
-$(app_test_objects): $(moose_config_symlink)
-$(excluded_objects): $(moose_config_symlink)
-
-else
-
-$(app_objects): $(moose_config)
-$(app_test_objects): $(moose_config)
-$(excluded_objects): $(moose_config_symlink)
-
-endif
-
 # plugin files
 plugfiles   := $(shell find $(PLUGIN_DIR) -regex "[^\#~]*\.C" 2>/dev/null)
 cplugfiles  := $(shell find $(PLUGIN_DIR) -name "*.c" 2>/dev/null)
@@ -174,6 +160,9 @@ plugins	    += $(patsubst %.f90, %-$(METHOD).plugin, $(f90plugfiles))
 MAIN_DIR    ?= $(APPLICATION_DIR)/src
 main_src    := $(MAIN_DIR)/main.C
 main_object := $(patsubst %.C, %.$(obj-suffix), $(main_src))
+
+# all objects that make up an application
+all_app_objects := $(app_objects) $(app_test_objects) $(plugins) $(main_object) $(excluded_objects)
 
 # dependency files
 app_deps     := $(patsubst %.$(obj-suffix), %.$(obj-suffix).d, $(objects)) \
@@ -219,7 +208,7 @@ link_names := $(foreach i, $(include_files), $(all_header_dir)/$(notdir $(i)))
 $(eval $(call all_header_dir_rule, $(all_header_dir)))
 $(call symlink_rules, $(all_header_dir), $(include_files))
 
-header_symlinks: $(all_header_dir) $(link_names)
+$(APPLICATION_NAME)_header_symlinks: $(all_header_dir) $(link_names)
 app_INCLUDE = -I$(all_header_dir)
 
 else # No Header Symlinks
@@ -291,6 +280,20 @@ LIBRARY_SUFFIX :=
 #
 ###############################################################################
 
+ifeq ($(MOOSE_HEADER_SYMLINKS),true)
+
+# If we are compiling with header symlinks, we don't want to start compiling any
+# object files until all symlinking is completed. The first dependency in the
+# list below ensures this.
+
+$(all_app_objects) : | $(APPLICATION_NAME)_header_symlinks $(moose_config_symlink)
+
+else
+
+$(all_app_objects) : $(moose_config)
+
+endif
+
 # Instantiate a new suffix rule for the module loader
 $(eval $(call CXX_RULE_TEMPLATE,_with$(app_LIB_SUFFIX)))
 
@@ -335,7 +338,7 @@ endif
 $(app_HEADER): curr_dir    := $(APPLICATION_DIR)
 $(app_HEADER): curr_app    := $(APPLICATION_NAME)
 $(app_HEADER): all_header_dir := $(all_header_dir)
-$(app_HEADER): $(app_HEADER_deps)
+$(app_HEADER): $(app_HEADER_deps) | $(all_header_dir)
 	@echo "Checking if header needs updating: "$@"..."
 	$(shell $(FRAMEWORK_DIR)/scripts/get_repo_revision.py $(curr_dir) $@ $(curr_app))
 	@ln -sf $@ $(all_header_dir)
@@ -414,7 +417,7 @@ endif
 
 # Write resource file
 $(app_resource):
-	@echo "Creating resource file"
+	@echo "Creating Resource file $@"
 	@$(shell $(FRAMEWORK_DIR)/scripts/write_appresource_file.py $(app_resource) $(APPLICATION_NAME) \
      $(libmesh_CXXFLAGS) \
      compiler_type=$(compilertype) \
@@ -423,7 +426,7 @@ $(app_resource):
 
 # Update and Copy resource file to prefix/bin
 install_$(APPLICATION_NAME)_resource:
-	@echo "Installing $(APPLICATION_NAME).yaml resource file"
+	@echo "Installing $(APPLICATION_NAME).yaml Resource file"
 	@$(shell $(FRAMEWORK_DIR)/scripts/write_appresource_file.py $(app_resource) $(APPLICATION_NAME) installation_type=relocated)
 	@mkdir -p $(bin_install_dir)
 	@cp $(app_resource) $(bin_install_dir)

--- a/framework/build.mk
+++ b/framework/build.mk
@@ -110,11 +110,7 @@ endif
 
 all:
 
-# Add all header symlinks as dependencies to this target
-header_symlinks:
-
 unity_files:
-
 
 #
 # C++ rules
@@ -312,8 +308,9 @@ endif
 # out to be more trouble than it was worth to get working.
 #
 PLUGIN_FLAGS := -shared -fPIC -Wl,-undefined,dynamic_lookup
+
+# we add include/base so that MooseConfig.h can be found, which is absent from the symlink dirs
 %-$(METHOD).plugin : %.C
-	# we add include/base so that MooseConfig.h can be found, which is absent from the symlink dirs
 	@$(libmesh_CXX) $(libmesh_CPPFLAGS) $(ADDITIONAL_CPPFLAGS) $(CXXFLAGS) $(libmesh_CXXFLAGS) $(PLUGIN_FLAGS) $(app_INCLUDES) $(libmesh_INCLUDE) -I $(FRAMEWORK_DIR)/include/base $< -o $@
 %-$(METHOD).plugin : %.c
 	@echo "Compiling C Plugin (in "$(METHOD)" mode) "$<"..."

--- a/framework/moose.mk
+++ b/framework/moose.mk
@@ -199,7 +199,7 @@ $(moose_config_symlink): $(moose_config) | $(moose_all_header_dir)
 	@echo "Symlinking MOOSE configure "$(moose_config_symlink)
 	@ln -sf $(moose_config) $(moose_config_symlink)
 
-header_symlinks: $(all_header_dir) $(link_names)
+moose_header_symlinks: $(all_header_dir) $(link_names)
 moose_INC_DIRS := $(all_header_dir)
 
 else # No Header Symlinks
@@ -329,7 +329,7 @@ app_DIRS     := $(FRAMEWORK_DIR)
 
 moose_revision_header := $(FRAMEWORK_DIR)/include/base/MooseRevision.h
 
-all: libmesh_submodule_status header_symlinks $(moose_revision_header) moose
+all: libmesh_submodule_status moose
 
 # revision header
 moose_GIT_DIR := $(shell cd "$(FRAMEWORK_DIR)" && which git &> /dev/null && git rev-parse --show-toplevel)
@@ -341,7 +341,7 @@ ifeq (x$(moose_HEADER_deps),x)
   moose_HEADER_deps := $(realpath $(moose_GIT_DIR)/HEAD $(moose_GIT_DIR)/index)
 endif
 
-$(moose_revision_header): $(moose_HEADER_deps)
+$(moose_revision_header): $(moose_HEADER_deps) | $(moose_all_header_dir)
 	@echo "Checking if header needs updating: "$@"..."
 	$(shell $(FRAMEWORK_DIR)/scripts/get_repo_revision.py $(FRAMEWORK_DIR) \
 	  $(moose_revision_header) MOOSE)
@@ -363,7 +363,7 @@ endif
 libmesh_submodule_status:
 	@if [ x$(libmesh_message) != "x" ]; then printf $(libmesh_message); fi
 
-moose: $(moose_LIB)
+moose: $(moose_revision_header) $(moose_LIB)
 
 # [JWP] With libtool, there is only one link command, it should work whether you are creating
 # shared or static libraries, and it should be portable across Linux and Mac...
@@ -393,7 +393,8 @@ $(moose_LIB): $(moose_objects) $(pcre_LIB) $(gtest_LIB) $(hit_LIB) $(pyhit_LIB)
 
 ifeq ($(MOOSE_HEADER_SYMLINKS),true)
 
-$(moose_objects): $(moose_config_symlink)
+
+$(moose_objects): $(moose_config_symlink) | moose_header_symlinks
 
 else
 

